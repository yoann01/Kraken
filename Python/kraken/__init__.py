"""Kraken Framework."""

import logging
import os

__all__ = ['core', 'helpers', 'plugins', 'ui']


krakenPath = os.path.abspath(os.path.join(os.path.dirname(os.path.realpath(__file__)), '..', '..'))
if os.environ.get('KRAKEN_PATH', None) is None:
    os.environ['KRAKEN_PATH'] = krakenPath

krakenExtsPath = os.path.join(krakenPath, 'Exts')
if 'FABRIC_EXTS_PATH' in os.environ:
    if krakenExtsPath not in os.environ['FABRIC_EXTS_PATH']:
        os.environ['FABRIC_EXTS_PATH'] = krakenExtsPath + os.pathsep + os.environ['FABRIC_EXTS_PATH']

krakenPresetsPath = os.path.join(krakenPath, 'Presets', 'DFG')
if 'FABRIC_DFG_PATH' in os.environ:
    if krakenPresetsPath not in os.environ['FABRIC_DFG_PATH']:
        os.environ['FABRIC_DFG_PATH'] = krakenPresetsPath + os.pathsep + os.environ['FABRIC_DFG_PATH']


logging.basicConfig(format='[KRAKEN] %(levelname)s: %(message)s', level=logging.INFO)

# Custom inform level for use with UI label getting added to the status bar.
logging.INFORM = 25
logging.addLevelName(logging.INFORM, 'INFORM')
logging.Logger.inform = lambda inst, msg, *args, **kwargs: inst.log(logging.INFORM, msg, *args, **kwargs)


def release():
    print "RELEASE THE KRAKEN!!!!"
