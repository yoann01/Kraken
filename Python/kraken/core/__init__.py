"""Kraken Core."""

VERSION_MAJOR = 1
VERSION_MINOR = 2
VERSION_BUILD = 6
VERSION_SUFFIX = ""


def getVersion():
    """Contatenates the version globals and returns the current version of
    Kraken.

    Returns:
        str: Current version of Kraken.

    """

    versionString = str(VERSION_MAJOR) + "." + str(VERSION_MINOR) + "." + str(VERSION_BUILD)
    if VERSION_SUFFIX:
        versionString = versionString + "-" + VERSION_SUFFIX

    return versionString
