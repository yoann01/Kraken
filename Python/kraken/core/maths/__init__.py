"""Kraken - math module."""

from math_object import MathObject
from vec2 import Vec2
from vec3 import Vec3
from vec4 import Vec4
from quat import Quat
from euler import Euler
from xfo import Xfo
from mat33 import Mat33
from mat44 import Mat44
from rotation_order import RotationOrder
from color import Color
from constants import DEG_TO_RAD, RAD_TO_DEG, PI


def Math_radToDeg(val):
    """Converts radians to degrees.

    Args:
        val (float): value to convert to degrees.

    Returns:
        float: Value in degrees.

    """

    return val * RAD_TO_DEG

def Math_degToRad(val):
    """Converts degrees to radians.

    Args:
        val (float): value to convert to radians.

    Returns:
        float: Value in radians.

    """

    return val * DEG_TO_RAD


def decodeValue(jsonData):
    """Returns a constructed math value based on the provided json data.

    Args:
        jsondata (dict): The JSON data to use to decode into a Math value.

    Returns:
        object: The constructed math value

    """

    if type(jsonData) is not dict:
        return jsonData

    if '__mathObjectClass__' not in jsonData:
        raise Exception("Invalid JSON data for constructing value:" + str(jsonData))

    if jsonData['__mathObjectClass__'] == 'Vec2':
        val = Vec2()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec3':
        val = Vec3()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Vec4':
        val = Vec4()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Euler':
        val = Euler()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Quat':
        val = Quat()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Xfo':
        val = Xfo()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat33':
        val = Mat33()
        val.jsonDecode(jsonData, decodeValue)
    elif jsonData['__mathObjectClass__'] == 'Mat44':
        val = Mat44()
        val.jsonDecode(jsonData, decodeValue)
    else:
        raise Exception("Unsupported Math type:" + jsonData['__mathObjectClass__'])

    return val
