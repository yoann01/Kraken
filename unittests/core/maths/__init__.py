
import unittest

import test_vec2
import test_vec3
import test_vec4
import test_color
import test_euler
import test_mat33
import test_mat44
import test_quat
import test_rotation_order
import test_xfo

loadVec2Suite = test_vec2.suite()
loadVec3Suite = test_vec3.suite()
loadVec4Suite = test_vec4.suite()
loadColorSuite = test_color.suite()
loadEulerSuite = test_euler.suite()
loadMat33Suite = test_mat33.suite()
loadMat44Suite = test_mat44.suite()
loadQuatSuite = test_quat.suite()
loadRotationOrderSuite = test_rotation_order.suite()
loadXfoSuite = test_xfo.suite()


def suite():
    suites = [
        loadVec2Suite,
        loadVec3Suite,
        loadVec4Suite,
        loadColorSuite,
        loadEulerSuite,
        loadMat33Suite,
        loadMat44Suite,
        loadQuatSuite,
        loadRotationOrderSuite,
        loadXfoSuite]

    return unittest.TestSuite(suites)


if __name__ == '__main__':
    runner = unittest.TextTestRunner()
    runner.run(suite())
