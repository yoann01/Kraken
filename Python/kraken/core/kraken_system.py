"""KrakenSystem - objects.kraken_core module.

Classes:
KrakenSystem - Class for constructing the Fabric Engine Core client.

"""

import logging
import os
import sys
import json
import importlib
from collections import OrderedDict

import FabricEngine.Core

# import kraken
from kraken.core.profiler import Profiler
from kraken.plugins import getFabricClient

from kraken.log import getLogger
from kraken.log.utils import fabricCallback

logger = getLogger('kraken')


class KrakenSystem(object):
    """The KrakenSystem is a singleton object used to provide an interface with
    the FabricEngine Core and RTVal system."""

    __instance = None

    def __init__(self):
        """Initializes the Kraken System object."""

        super(KrakenSystem, self).__init__()

        self.client = None
        self.typeDescs = None
        self.registeredTypes = None
        self.loadedExtensions = []

        self.registeredConfigs = OrderedDict()
        self.registeredComponents = OrderedDict()
        # self.moduleImportManager = ModuleImportManager()


    def loadCoreClient(self):
        """Loads the Fabric Engine Core Client"""

        if self.client is None:
            Profiler.getInstance().push("loadCoreClient")

            client = getFabricClient()
            if client is None:
                options = {
                    'reportCallback': fabricCallback,
                    'guarded': True
                }

                client = FabricEngine.Core.createClient(options)

            self.client = client

            self.loadExtension('Math')
            self.loadExtension('Kraken')
            self.loadExtension('KrakenForCanvas')

            Profiler.getInstance().pop()

    def getCoreClient(self):
        """Returns the Fabric Engine Core Client owned by the KrakenSystem

        Returns:
            object: The Fabric Engine Core Client

        """

        if self.client is None:
            self.loadCoreClient()

        return self.client

    def loadExtension(self, extension):
        """Loads the given extension and updates the registeredTypes cache.

        Args:
            extension (str): The name of the extension to load.

        """

        if extension not in self.loadedExtensions:
            Profiler.getInstance().push("loadExtension:" + extension)
            self.client.loadExtension(extension)
            self.registeredTypes = self.client.RT.types
            self.typeDescs = self.client.RT.getRegisteredTypes()
            # Cache the loaded extension so that we aviod refreshing the typeDescs cache(costly)
            self.loadedExtensions.append(extension)
            Profiler.getInstance().pop()


    # ==============
    # RTVal Methods
    # ==============
    def convertFromRTVal(self, target, RTTypeName=None):
        """Generates an RTVal object based on the simple type of target
        and passes target to constructor. Converts a property of an RTVal object
        to its own pytholn RTVal object


        Args:
            target (RTVal): The RTVal object or property to cast
            RTTypeName (str): The type of RTVal to convert to

        Returns:
            RTVal: The RTVal object

        """

        self.loadCoreClient()

        if RTTypeName is None:
            RTTypeName = target.type('String').getSimpleType()
        rtValType = getattr(self.client.RT.types, RTTypeName)
        pythonRTVal = rtValType(target)

        return pythonRTVal

    def constructRTVal(self, dataType, defaultValue=None):
        """Constructs a new RTVal using the given name and optional devault value.

        Args:
            dataType (str): The name of the data type to construct.
            defaultValue (value): The default value to use to initialize the RTVal

        Returns:
            object: The constructed RTval.

        """

        self.loadCoreClient()
        klType = getattr(self.registeredTypes, dataType)

        if defaultValue is not None:
            if hasattr(defaultValue, '_rtval'):
                return defaultValue._rtval

            typeDesc = self.typeDescs[dataType]
            if 'members' in typeDesc:
                try:
                    value = klType.create()
                except:
                    try:
                        return klType()
                    except Exception as e:
                        raise Exception("Error constructing RTVal:" + dataType)

                for i in xrange(0, len(typeDesc['members'])):
                    memberName = typeDesc['members'][i]['name']
                    memberType = typeDesc['members'][i]['type']
                    if memberName in defaultValue:
                        setattr(value, memberName, self.constructRTVal(memberType, getattr(defaultValue, memberName)))

                return value

            else:
                return klType(defaultValue)
        else:
            try:
                return klType.create()
            except:
                try:
                    return klType()
                except Exception as e:
                    raise Exception("Error constructing RTVal:" + dataType)

    def rtVal(self, dataType, defaultValue=None):
        """Constructs a new RTVal using the given name and optional devault value.

        Args:
            dataType (str): The name of the data type to construct.
            defaultValue (value): The default value to use to initialize the RTVal

        Returns:
            object: The constructed RTval.

        """

        return self.constructRTVal(dataType, defaultValue)

    def isRTVal(self, value):
        """Returns true if the given value is an RTVal.

        Args:
            value (value): value to test.

        Returns:
            bool: True if successful.

        """

        return str(type(value)) == "<type 'PyRTValObject'>"

    def getRTValTypeName(self, rtval):
        """Returns the name of the type, handling extracting the name from KL RTVals.

        Args:
            rtval (rtval): The rtval to extract the name from.

        Returns:
            bool: True if successful.

        """

        if ks.isRTVal(rtval):
            return json.loads(rtval.type("Type").jsonDesc("String").getSimpleType())['name']
        else:
            return "None"

    # ==================
    # Config Methods
    # ==================
    def registerConfig(self, configClass):
        """Registers a config Python class with the KrakenSystem so ti can be built by the rig builder.

        Args:
            configClass (str): The Python class of the config

        """

        configModulePath = configClass.__module__ + "." + configClass.__name__

        self.registeredConfigs[configModulePath] = configClass

    def getConfigClass(self, className):
        """Returns the registered Python config class with the given name

        Args:
            className (str): The name of the Python config class

        Returns:
            object: The Python config class

        """

        if className not in self.registeredConfigs:
            raise Exception("Config with that class not registered:" + className)

        return self.registeredConfigs[className]

    def getConfigClassNames(self):
        """Returns the names of the registered Python config classes

        Returns:
            list: The array of config class names.

        """

        return self.registeredConfigs.keys()

    # ==================
    # Component Methods
    # ==================
    def registerComponent(self, componentClass):
        """Registers a component Python class with the KrakenSystem so ti can be built by the rig builder.

        Args:
            componentClass (str): The Python class of the component

        """
        componentClassPath = componentClass.__module__ + "." + componentClass.__name__
        if componentClassPath in self.registeredComponents:
            # we allow reregistring of components because as a component's class is edited
            # it will be re-imported by python(in Maya), and the classes reregistered.
            pass

        self.registeredComponents[componentClassPath] = componentClass

    def getComponentClass(self, className):
        """Returns the registered Python component class with the given name

        Args:
            className (str): The name of the Python component class

        Returns:
            object: The Python component class

        """

        if className not in self.registeredComponents:
            raise Exception("Component with that class not registered:" + className)

        return self.registeredComponents[className]

    def getComponentClassNames(self):
        """Returns the names of the registered Python component classes

        Returns:
            list: The array of component class names.

        """

        return self.registeredComponents.keys()

    def loadComponentModules(self):
        """Loads all the component modules and configs specified in the 'KRAKEN_PATHS' environment variable.

        The kraken_components are loaded at all times.

        Returns:
            bool: True if all components loaded, else False.

        """

        for componentClassPath in self.registeredComponents:
            componentModulePath = self.registeredComponents[componentClassPath].__module__
            if componentModulePath in sys.modules:
                del(sys.modules[componentModulePath])

        self.registeredComponents = {}

        logger.info("Loading component modules...")

        def __importDirRecursive(path, parentModulePath=''):
            isSuccessful = True

            contents = os.listdir(path)
            moduleFilefound = False
            for item in contents:
                if os.path.isfile(os.path.join(path, item)):
                    if item == "__init__.py":
                        if parentModulePath == '':
                            modulePath = os.path.basename(path)

                            moduleParentFolder = os.path.split( path )[0]
                            if moduleParentFolder not in sys.path:
                                sys.path.append(moduleParentFolder)
                        else:
                            modulePath = parentModulePath + '.' + os.path.basename(path)
                        moduleFilefound = True


            if moduleFilefound:
                logger.info(" " + path + ":")
                for i, item in enumerate(contents):
                    if os.path.isfile(os.path.join(path, item)):

                        # Parse all the files of given path and import python
                        # modules. The files in these folders really should be
                        # limited to components, otherwise we are loading more
                        # than modules and that is not clear.

                        # TODO: Figure out a way to limit imports to just rig
                        # component modules.

                        if item.endswith(".py") and item != "__init__.py":
                            module = modulePath + "." + item[:-3]
                            try:
                                logger.info("  " + module)
                                importlib.import_module(module)

                            except ImportError, e:
                                isSuccessful = False
                                logging.exception("Error importing '" + module)

                            except Exception, e:
                                isSuccessful = False
                                logging.exception("Error Loading Modules'" + module)

                logging.info("")


            for item in contents:
                if os.path.isdir(os.path.join(path, item)):
                    if moduleFilefound:
                        if not __importDirRecursive(os.path.join(path, item), modulePath):
                            isSuccessful = False
                    else:
                        if not __importDirRecursive(os.path.join(path, item)):
                            isSuccessful = False

            return isSuccessful


        # Find the kraken examples module in the same folder as the kraken module.
        default_component_path = os.path.normpath(os.path.join(os.environ.get('KRAKEN_PATH'), 'Python', 'kraken_components'))
        isSuccessful = __importDirRecursive(default_component_path)

        pathsVar = os.getenv('KRAKEN_PATHS')
        if pathsVar is not None:
            pathsList = pathsVar.split(os.pathsep)
            for path in pathsList:

                if path == '':
                    continue

                if not os.path.exists(path):
                    logging.info("Invalid Kraken Path: " + path)
                    continue

                if not __importDirRecursive(path):
                    isSuccessful = False

        return isSuccessful


    @classmethod
    def getInstance(cls):
        """This class method returns the singleton instance for the KrakenSystem

        Returns:
            object: The singleton instance.

        """

        if cls.__instance is None:
            cls.__instance = KrakenSystem()

        return cls.__instance


ks = KrakenSystem.getInstance()
