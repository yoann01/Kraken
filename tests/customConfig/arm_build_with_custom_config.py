
from kraken import plugins
from kraken.core.maths import Xfo, Vec3, Quat
from kraken.core.configs.config import Config

from kraken_components.biped.arm_component import ArmComponentGuide, ArmComponentRig


class CustomConfig(Config):
    """Base Configuration for Kraken builders."""

    def __init__(self):
        super(CustomConfig, self).__init__()

    # ======================
    # Name Template Methods
    # ======================
    def initNameTemplate(self):
        """Initializes the name template.

        Returns:
            dict: name template.

        """

        nameTemplate = super(CustomConfig, self).initNameTemplate()
        nameTemplate["formats"] = {
            "Container": ["name"],
            "Layer": ["container", "sep", "name"],
            "ComponentGroup": ["location", "sep", "name", "sep", "type"],
            "default": ["location", "sep", "component", "sep", "name", "sep", "type"],
        }

        return nameTemplate




armGuide = ArmComponentGuide("arm")
armGuide.loadData({
    "name": "Arm",
    "location": "L",
    "bicepXfo": Xfo(Vec3(2.27, 15.295, -0.753)),
    "forearmXfo": Xfo(Vec3(5.039, 13.56, -0.859)),
    "wristXfo": Xfo(Vec3(7.1886, 12.2819, 0.4906)),
    "handXfo": Xfo(tr=Vec3(7.1886, 12.2819, 0.4906),
                   ori=Quat(Vec3(-0.0865, -0.2301, -0.2623), 0.9331)),
    "bicepFKCtrlSize": 1.75,
    "forearmFKCtrlSize": 1.5
})

armGuideData = armGuide.getRigBuildData()

arm = ArmComponentRig()
arm.loadData(armGuideData)

builder = plugins.getBuilder()
builder.build(arm)

# Set the custom config as the new singleton value.
CustomConfig.makeCurrent()

builder = plugins.getBuilder()
builder.build(arm)
