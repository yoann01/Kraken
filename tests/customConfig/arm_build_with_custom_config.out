Failed to find DCC builder. Falling back to Python builder.
building: controls.Arm as: Arm_L_cmp type: ComponentGroup
buildGroup: controls.Arm as: Arm_L_cmp
building: controls.Arm.inputs as: Arm_L_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.inputs as: Arm_L_inputs_hrc
building: controls.Arm.inputs.globalSRT as: Arm_L_globalSRT_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.globalSRT as: Arm_L_globalSRT_cmpIn
building: controls.Arm.inputs.root as: Arm_L_root_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.root as: Arm_L_root_cmpIn
building: controls.Arm.outputs as: Arm_L_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.outputs as: Arm_L_outputs_hrc
building: controls.Arm.outputs.bicep as: Arm_L_bicep_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.bicep as: Arm_L_bicep_cmpOut
building: controls.Arm.outputs.forearm as: Arm_L_forearm_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.forearm as: Arm_L_forearm_cmpOut
building: controls.Arm.outputs.wrist as: Arm_L_wrist_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.wrist as: Arm_L_wrist_cmpOut
building: controls.Arm.bicepFK as: Arm_L_bicepFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK as: Arm_L_bicepFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK as: Arm_L_bicepFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK as: Arm_L_bicepFK_ctrl
building: controls.Arm.bicepFK.bicepFK.forearmFK as: Arm_L_forearmFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK.bicepFK.forearmFK as: Arm_L_forearmFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: Arm_L_forearmFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: Arm_L_forearmFK_ctrl
building: controls.Arm.IK as: Arm_L_IK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.IK as: Arm_L_IK_ctrlSpace
building: controls.Arm.IK.IK as: Arm_L_IK_ctrl type: Control
buildControl:controls.Arm.IK.IK as: Arm_L_IK_ctrl
building: controls.Arm.ikPosition as: Arm_L_ikPosition_null type: Transform
buildGroup: controls.Arm.ikPosition as: Arm_L_ikPosition_null
building: controls.Arm.UpV as: Arm_L_UpV_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.UpV as: Arm_L_UpV_ctrlSpace
building: controls.Arm.UpV.UpV as: Arm_L_UpV_ctrl type: Control
buildControl:controls.Arm.UpV.UpV as: Arm_L_UpV_ctrl
building: deformers as: deformers type: Layer
buildLayer: deformers as: deformers
building: deformers.Arm as: Arm_L_cmp type: ComponentGroup
buildGroup: deformers.Arm as: Arm_L_cmp
building: deformers.Arm.bicep as: Arm_L_bicep_def type: Joint
buildJoint: deformers.Arm.bicep as: Arm_L_bicep_def
building: deformers.Arm.forearm as: Arm_L_forearm_def type: Joint
buildJoint: deformers.Arm.forearm as: Arm_L_forearm_def
building: deformers.Arm.wrist as: Arm_L_wrist_def type: Joint
buildJoint: deformers.Arm.wrist as: Arm_L_wrist_def
building: controls as: controls type: Layer
buildLayer: controls as: controls
building: controls.Arm as: Arm_L_cmp type: ComponentGroup
buildGroup: controls.Arm as: Arm_L_cmp
building: controls.Arm.inputs as: Arm_L_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.inputs as: Arm_L_inputs_hrc
building: controls.Arm.inputs.globalSRT as: Arm_L_globalSRT_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.globalSRT as: Arm_L_globalSRT_cmpIn
building: controls.Arm.inputs.root as: Arm_L_root_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.root as: Arm_L_root_cmpIn
building: controls.Arm.outputs as: Arm_L_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.outputs as: Arm_L_outputs_hrc
building: controls.Arm.outputs.bicep as: Arm_L_bicep_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.bicep as: Arm_L_bicep_cmpOut
building: controls.Arm.outputs.forearm as: Arm_L_forearm_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.forearm as: Arm_L_forearm_cmpOut
building: controls.Arm.outputs.wrist as: Arm_L_wrist_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.wrist as: Arm_L_wrist_cmpOut
building: controls.Arm.bicepFK as: Arm_L_bicepFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK as: Arm_L_bicepFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK as: Arm_L_bicepFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK as: Arm_L_bicepFK_ctrl
building: controls.Arm.bicepFK.bicepFK.forearmFK as: Arm_L_forearmFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK.bicepFK.forearmFK as: Arm_L_forearmFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: Arm_L_forearmFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: Arm_L_forearmFK_ctrl
building: controls.Arm.IK as: Arm_L_IK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.IK as: Arm_L_IK_ctrlSpace
building: controls.Arm.IK.IK as: Arm_L_IK_ctrl type: Control
buildControl:controls.Arm.IK.IK as: Arm_L_IK_ctrl
building: controls.Arm.ikPosition as: Arm_L_ikPosition_null type: Transform
buildGroup: controls.Arm.ikPosition as: Arm_L_ikPosition_null
building: controls.Arm.UpV as: Arm_L_UpV_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.UpV as: Arm_L_UpV_ctrlSpace
building: controls.Arm.UpV.UpV as: Arm_L_UpV_ctrl type: Control
buildControl:controls.Arm.UpV.UpV as: Arm_L_UpV_ctrl
building: deformers.Arm as: Arm_L_cmp type: ComponentGroup
buildGroup: deformers.Arm as: Arm_L_cmp
building: deformers.Arm.bicep as: Arm_L_bicep_def type: Joint
buildJoint: deformers.Arm.bicep as: Arm_L_bicep_def
building: deformers.Arm.forearm as: Arm_L_forearm_def type: Joint
buildJoint: deformers.Arm.forearm as: Arm_L_forearm_def
building: deformers.Arm.wrist as: Arm_L_wrist_def type: Joint
buildJoint: deformers.Arm.wrist as: Arm_L_wrist_def
Failed to find DCC builder. Falling back to Python builder.
building: controls.Arm as: L_Arm_cmp type: ComponentGroup
buildGroup: controls.Arm as: L_Arm_cmp
building: controls.Arm.inputs as: L_Arm_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.inputs as: L_Arm_inputs_hrc
building: controls.Arm.inputs.globalSRT as: L_Arm_globalSRT_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.globalSRT as: L_Arm_globalSRT_cmpIn
building: controls.Arm.inputs.root as: L_Arm_root_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.root as: L_Arm_root_cmpIn
building: controls.Arm.outputs as: L_Arm_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.outputs as: L_Arm_outputs_hrc
building: controls.Arm.outputs.bicep as: L_Arm_bicep_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.bicep as: L_Arm_bicep_cmpOut
building: controls.Arm.outputs.forearm as: L_Arm_forearm_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.forearm as: L_Arm_forearm_cmpOut
building: controls.Arm.outputs.wrist as: L_Arm_wrist_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.wrist as: L_Arm_wrist_cmpOut
building: controls.Arm.bicepFK as: L_Arm_bicepFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK as: L_Arm_bicepFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK as: L_Arm_bicepFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK as: L_Arm_bicepFK_ctrl
building: controls.Arm.bicepFK.bicepFK.forearmFK as: L_Arm_forearmFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK.bicepFK.forearmFK as: L_Arm_forearmFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: L_Arm_forearmFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: L_Arm_forearmFK_ctrl
building: controls.Arm.IK as: L_Arm_IK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.IK as: L_Arm_IK_ctrlSpace
building: controls.Arm.IK.IK as: L_Arm_IK_ctrl type: Control
buildControl:controls.Arm.IK.IK as: L_Arm_IK_ctrl
building: controls.Arm.ikPosition as: L_Arm_ikPosition_null type: Transform
buildGroup: controls.Arm.ikPosition as: L_Arm_ikPosition_null
building: controls.Arm.UpV as: L_Arm_UpV_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.UpV as: L_Arm_UpV_ctrlSpace
building: controls.Arm.UpV.UpV as: L_Arm_UpV_ctrl type: Control
buildControl:controls.Arm.UpV.UpV as: L_Arm_UpV_ctrl
building: deformers as: deformers type: Layer
buildLayer: deformers as: deformers
building: deformers.Arm as: L_Arm_cmp type: ComponentGroup
buildGroup: deformers.Arm as: L_Arm_cmp
building: deformers.Arm.bicep as: L_Arm_bicep_def type: Joint
buildJoint: deformers.Arm.bicep as: L_Arm_bicep_def
building: deformers.Arm.forearm as: L_Arm_forearm_def type: Joint
buildJoint: deformers.Arm.forearm as: L_Arm_forearm_def
building: deformers.Arm.wrist as: L_Arm_wrist_def type: Joint
buildJoint: deformers.Arm.wrist as: L_Arm_wrist_def
building: controls as: controls type: Layer
buildLayer: controls as: controls
building: controls.Arm as: L_Arm_cmp type: ComponentGroup
buildGroup: controls.Arm as: L_Arm_cmp
building: controls.Arm.inputs as: L_Arm_inputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.inputs as: L_Arm_inputs_hrc
building: controls.Arm.inputs.globalSRT as: L_Arm_globalSRT_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.globalSRT as: L_Arm_globalSRT_cmpIn
building: controls.Arm.inputs.root as: L_Arm_root_cmpIn type: ComponentInput
buildLocator: controls.Arm.inputs.root as: L_Arm_root_cmpIn
building: controls.Arm.outputs as: L_Arm_outputs_hrc type: HierarchyGroup
buildHierarchyGroup: controls.Arm.outputs as: L_Arm_outputs_hrc
building: controls.Arm.outputs.bicep as: L_Arm_bicep_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.bicep as: L_Arm_bicep_cmpOut
building: controls.Arm.outputs.forearm as: L_Arm_forearm_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.forearm as: L_Arm_forearm_cmpOut
building: controls.Arm.outputs.wrist as: L_Arm_wrist_cmpOut type: ComponentOutput
buildLocator: controls.Arm.outputs.wrist as: L_Arm_wrist_cmpOut
building: controls.Arm.bicepFK as: L_Arm_bicepFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK as: L_Arm_bicepFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK as: L_Arm_bicepFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK as: L_Arm_bicepFK_ctrl
building: controls.Arm.bicepFK.bicepFK.forearmFK as: L_Arm_forearmFK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.bicepFK.bicepFK.forearmFK as: L_Arm_forearmFK_ctrlSpace
building: controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: L_Arm_forearmFK_ctrl type: Control
buildControl:controls.Arm.bicepFK.bicepFK.forearmFK.forearmFK as: L_Arm_forearmFK_ctrl
building: controls.Arm.IK as: L_Arm_IK_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.IK as: L_Arm_IK_ctrlSpace
building: controls.Arm.IK.IK as: L_Arm_IK_ctrl type: Control
buildControl:controls.Arm.IK.IK as: L_Arm_IK_ctrl
building: controls.Arm.ikPosition as: L_Arm_ikPosition_null type: Transform
buildGroup: controls.Arm.ikPosition as: L_Arm_ikPosition_null
building: controls.Arm.UpV as: L_Arm_UpV_ctrlSpace type: CtrlSpace
buildGroup: controls.Arm.UpV as: L_Arm_UpV_ctrlSpace
building: controls.Arm.UpV.UpV as: L_Arm_UpV_ctrl type: Control
buildControl:controls.Arm.UpV.UpV as: L_Arm_UpV_ctrl
building: deformers.Arm as: L_Arm_cmp type: ComponentGroup
buildGroup: deformers.Arm as: L_Arm_cmp
building: deformers.Arm.bicep as: L_Arm_bicep_def type: Joint
buildJoint: deformers.Arm.bicep as: L_Arm_bicep_def
building: deformers.Arm.forearm as: L_Arm_forearm_def type: Joint
buildJoint: deformers.Arm.forearm as: L_Arm_forearm_def
building: deformers.Arm.wrist as: L_Arm_wrist_def type: Joint
buildJoint: deformers.Arm.wrist as: L_Arm_wrist_def
